add sku option for sqldatabase

guyomog78 opened the pull request:
##### SUMMARY
Change the non functional **edition** option to a functional *sku* option.

Partial Fixes #153 *Unable to specify a SKU*
Fixes #126 
##### ISSUE TYPE
- Feature Pull Request
- Bugfix Pull Request

##### COMPONENT NAME
azure_rm_sqldatabase.py
azure_rm_sqldatabase_info.py

##### ADDITIONAL INFORMATION

The Sku option use the Python *Sku* class from azure.mgmt.sql.models to be able to push valide values.
The documentation is updated with the new SKU call option
The internal test to detect change is updated (replace **edition** by **sku**)

The sku entrie is then identical as the azure_rm_sqldatabase_info report

**example call before change**
```
  - name: Create (or update) SQL Database with SKU
    azure_rm_sqldatabase:
      resource_group: myResourceGroup
      server_name: sqlcrudtest-5961
      name: testdb
      location: eastus
      edition: standard
```

**example call after change**
```
  - name: Create (or update) SQL Database with SKU
    azure_rm_sqldatabase:
      resource_group: myResourceGroup
      server_name: sqlcrudtest-5961
      name: testdb
      location: eastus
      sku:
        name: S0
```
or with more options
**example call after change**
```
  - name: Create (or update) SQL Database with SKU
    azure_rm_sqldatabase:
      resource_group: myResourceGroup
      server_name: sqlcrudtest-5961
      name: testdb
      location: eastus
      sku:
        name: S0
        tier: Standard
        capacity: 50
```

**Attention** All option under sku should be coherent with the name value else an error is reported

guyomog78 committed:
Add SKU

start SKU minimum implementation
Functionnal with basic and standard need to improv for S0, S1 ...

guyomog78 committed:
Add functionnal SKU Choice

Replace the old non functionnal option **Edition**
by the **SKU** Dict

guyomog78 committed:
Extend sku availabe parameters

guyomog78 committed:
Merge branch 'dev' into fix-SQLSKU

Fred-sun commented on plugins/modules/azure_rm_sqldatabase.py:
```diff
@@ -77,23 +77,23 @@
             - Required if I(create_mode=restore_long_term_retention_backup), then this value is required.
             - Specifies the resource ID of the recovery point to restore from.
-    edition:
+    sku:
```
Argument 'size' and 'family' found in sku is listed in the argument_spec, but not documented in the module documentation

guyomog78 replied:
I have added missing arguments in the documentation

Fred-sun commented on plugins/modules/azure_rm_sqldatabase.py:
```diff
@@ -334,8 +343,9 @@ def exec_module(self, **kwargs):
-                    self.parameters["edition"] = _snake_to_camel(kwargs[key], True)
+                elif key == "sku":
+                    ev = kwargs[key]
+                    self.parameters["sku"] = Sku(name = ev['name'], tier = ev['tier'], size = ev['size'], family = ev['family'], capacity = ev['capacity'] )
```
```suggestion
                    self.parameters["sku"] = Sku(name=ev['name'], tier=ev['tier'], size=ev['size'], family=ev['family'], capacity=ev['capacity'])
```

Fred-sun commented on plugins/modules/azure_rm_sqldatabase.py:
```diff
@@ -383,9 +393,9 @@ def exec_module(self, **kwargs):
-                        (self.parameters['edition'] != old_response['edition'])):
-                    self.to_do = Actions.Update
+                if (('sku' in self.parameters) and
+                         (self.parameters['sku'] != old_response['sku'])):
```
continuation line over-indented for visual indent

guyomog78 replied:
I use the same style usage already present. should I update my part?

Fred-sun commented on plugins/modules/azure_rm_sqldatabase.py:
```diff
@@ -383,9 +393,9 @@ def exec_module(self, **kwargs):
-                    self.to_do = Actions.Update
+                if (('sku' in self.parameters) and
+                         (self.parameters['sku'] != old_response['sku'])):
+                     self.to_do = Actions.Update
```
indentation is not a multiple of four and over-indented!

guyomog78 replied:
Ouch, shame on me! I correct It right now

guyomog78 committed:
Cosmetic "=" usage correction

Co-authored-by: Fred-sun <37327967+Fred-sun@users.noreply.github.com>

guyomog78 committed:
Fix PR Comment Returns

Fred-sun commented:
@guyomog78 Can you help tests/integration/the targets/azure_rm_sqlserver/tasks/main yml increase test cases? Thank you very much!

guyomog78 commented:
@Fred-sun Like make a test with the S0 mode and ensure that realy is a S0 mode, then change to P1 mode ensure that is well a P1?? someting like that?

guyomog78 commented:
OK, did you have an ETA in mind?

guyomog78 committed:
Improve Sqldatabase SKU return

Ensure that the sku name is name as requested
 - use current_service_objective_name instead of sku.name
 - sku.name is too generic
Add validation test for SKU usage

guyomog78 commented:
@Fred-sun, it was a little bit more difficulte as expected, but necessary to correct the sqldatabasinfo too to report the good SKU like documentation say.
when database information collected, sku name is equal to Standard for Service_objective_name S0,S1,S2.. Premium for Service_objective_name P1, P2 ,P3 ...
So, to use the expected information in sku.name (ex: S0, P2, DW100c , GP_FsV2_12...) I have used the current_service_objective_name retruned by the request.

To see current information: `Get-AzSqlServerServiceObjective -Location eastus`

guyomog78 committed:
Merge branch 'dev' into fix-SQLSKU

Fred-sun commented:
@guyomog78 Such parameter changes may cause the edition parameter previously configured by the user to be unavailable. Can the following two methods be used to avoid this problem?
First: Add Fail prompt. After this version, the edition parameter is changed to sku, and sku aliase is changed to sku.
Second: Keep the edition parameter, but set the default value and "mutually_exclusive=['edition', 'sku']", for example, when edition: Standard, the default sku.name=S1, and edition: premium, the default sku.name=P1.

guyomog78 commented:
@Fred-sun t it is a while that the edition parameter is not functionnal, but I understand your point of view.

I will try to keep for this time the edition parameters

guyomog78 committed:
re Add edition option

 - edition still present, but finally use arbitrary SKU
  - choice edition or Sku
  - update test to test edition too on first test
    - remember to remove next version

guyomog78 commented:
@Fred-sun It's Done!!
I have also used autopep8 and pylint in vscode and applied "format document"

On test file a have injected Edition value Premium and it's good. I have tested with standard too ;-)

et Voilà!

guyomog78 committed:
Merge branch 'dev' into fix-SQLSKU

Fred-sun commented on plugins/modules/azure_rm_sqldatabase.py:
```diff
@@ -298,6 +365,7 @@ def __init__(self):
                 choices=['present', 'absent']
             )
         )
+        mutually_exclusive = [['sky', 'edition']]
```
```suggestion
        mutually_exclusive = ['sku', 'edition']
```

guyomog78 replied:
Oups!. done

Fred-sun commented on plugins/modules/azure_rm_sqldatabase.py:
```diff
@@ -5,6 +5,8 @@
 # GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
 
 from __future__ import absolute_import, division, print_function
+from ansible_collections.azure.azcollection.plugins.module_utils.azure_rm_common import AzureRMModuleBase, format_resource_id
```
Import found before documentation variables

Fred-sun commented on plugins/modules/azure_rm_sqldatabase.py:
```diff
@@ -5,6 +5,8 @@
 
 from __future__ import absolute_import, division, print_function
+from ansible_collections.azure.azcollection.plugins.module_utils.azure_rm_common import AzureRMModuleBase, format_resource_id
+import time
```
Import found before documentation variables

Fred-sun committed:
Merge branch 'dev' into fix-SQLSKU

guyomog78 committed:
Fix import and exclusive

guyomog78 commented:
@Fred-sun Thanks a lot for review.
Now I'm aware about that the import part for script should be after documentation. It was modified by the "format document" under vscode. Now, I know that I need to revert this part for the next time if appens :-)

haiyuazhang committed:
Merge branch 'dev' into fix-SQLSKU
